fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/mixin/datagen/ModelMixin.java

```diff
@@ -0,0 +1,174 @@
+	@Unique
+	private final EnumMap<DisplayBuilder.Position, DisplayBuilder> displays = new EnumMap<>(DisplayBuilder.Position.class);
+	@Unique
+	private final List<ElementBuilder> elements = new ArrayList<>();
+	@Unique
+	private final List<OverrideBuilder> overrides = new ArrayList<>();
+	@Unique
+	private ItemModelBuilder.GuiLight guiLight = null;
+	@Unique
+	private boolean ambientOcclusion = true;
+
+	@Override
+	public Optional<Identifier> getParent() {
+		return parent;
+	}
+
+	@Override
+	public Set<TextureKey> getRequiredTextures() {
+		return requiredTextures;
+	}
+
+	@Override
+	public Optional<String> getVariant() {
+		return variant;
+	}
+
+	@Override
+	public Model withDisplay(DisplayBuilder.Position position, DisplayBuilder builder) {
+		this.displays.put(position, builder);
+		return (Model) (Object) this;
+	}
+
+	@Override
+	public EnumMap<DisplayBuilder.Position, DisplayBuilder> getDisplayBuilders() {
+		return displays;
+	}
+
+	@Override
+	public Model addElement(ElementBuilder builder) {
+		this.elements.add(builder);
+		return (Model) (Object) this;
+	}
+
+	@Override
+	public List<ElementBuilder> getElementBuilders() {
+		return elements;
+	}
+
+	@Override
+	public Model addOverride(OverrideBuilder builder) {
+		this.overrides.add(builder);
+		return (Model) (Object) this;
+	}
+
+	@Override
+	public List<OverrideBuilder> getOverrideBuilders() {
+		return overrides;
+	}
+
+	@Override
+	public Model setGuiLight(ItemModelBuilder.GuiLight guiLight) {
+		this.guiLight = guiLight;
+		return (Model) (Object) this;
+	}
+
+	@Override
+	public ItemModelBuilder.GuiLight getGuiLight() {
+		return guiLight;
+	}
+
+	@Override
+	public Model setAmbientOcclusion(boolean ambientOcclusion) {
+		this.ambientOcclusion = ambientOcclusion;
+		return (Model) (Object) this;
+	}
+
+	@Override
+	public boolean getAmbientOcclusion() {
+		return ambientOcclusion;
+	}
+
+	@Inject(method = "method_25851", at = @At("RETURN"), locals = LocalCapture.CAPTURE_FAILHARD)
+	public void addExtraProperties(Map<TextureKey, Identifier> map, CallbackInfoReturnable<JsonElement> cir, JsonObject jsonObject) {
+		if (!displays.isEmpty()) {
+			JsonObject display = new JsonObject();
+			this.displays.forEach((p, d) -> display.add(p.name().toLowerCase(), d.build()));
+			jsonObject.add("display", display);
+		}
+
+		if (!elements.isEmpty()) {
+			JsonArray elements = new JsonArray();
+			this.elements.forEach(e -> elements.add(e.build()));
+			jsonObject.add("elements", elements);
+		}
+
+		if (!overrides.isEmpty()) {
+			JsonArray overrides = new JsonArray();
+			this.overrides.forEach(o -> overrides.add(o.build()));
+			jsonObject.add("overrides", overrides);
+		}
+
+		if (guiLight != null) {
+			jsonObject.addProperty("gui_light", guiLight.name().toLowerCase());
+		}
+
+		if (!ambientOcclusion) {
+			jsonObject.addProperty("ambientocclusion", false);
+		}
+
+		displays.clear();
+		elements.clear();
+		overrides.clear();
+		guiLight = null;
+		ambientOcclusion = true;
```

Comment:
Why is this needed?

Reply:
Since these are mutable fields, any `Model` for which they're set will retain these properties even when it isn't intended. This was a concern at an earlier stage when standard models such as `Models.GENERATED` would retain these properties, though I'll try to give this a further look and see if it still applies.